implementations/micrometer-registry-opentelemetry/src/main/java/io/micrometer/opentelemetry/OpenTelemetryRegistry.java

```diff
@@ -0,0 +1,156 @@
+    }
+
+    @Override
+    protected Counter newCounter(Meter.Id id) {
+        DoubleCounter.Builder builder = otelMeter.doubleCounterBuilder(id.getName());
+        copyAttributes(id, builder);
+        if ( id.getBaseUnit() != null ) {
+            builder.setUnit(id.getBaseUnit());
+        }
+        return new OpenTelemetryCounter(id, builder.build());
+    }
+
+    @Override
+    protected Timer newTimer(Meter.Id id, DistributionStatisticConfig distributionStatisticConfig, PauseDetector pauseDetector) {
+        DoubleValueRecorder.Builder builder = otelMeter.doubleValueRecorderBuilder(id.getName())
+                .setUnit("s");
+        copyAttributes(id, builder);
+        return new OpenTelemetryTimer(id, clock, builder.build(), TimeUnit.SECONDS);
+    }
+
+    @Override
+    protected DistributionSummary newDistributionSummary(Meter.Id id, DistributionStatisticConfig distributionStatisticConfig, double scale) {
+        return null;
+    }
+
+    @Override
+    protected Meter newMeter(Meter.Id id, Meter.Type type, Iterable<Measurement> measurements) {
+        return new DefaultMeter(id, type, measurements);
+    }
+
+    @Override
+    protected <T> FunctionTimer newFunctionTimer(Meter.Id id, T obj, ToLongFunction<T> countFunction, ToDoubleFunction<T> totalTimeFunction, TimeUnit totalTimeFunctionUnit) {
+        // TODO: names must be different
+        DoubleSumObserver.Builder countObserver = otelMeter.doubleSumObserverBuilder(id.getName()+".count");
+        copyAttributes(id, countObserver);
+
+        DoubleValueObserver.Builder valueObserver = otelMeter.doubleValueObserverBuilder(id.getName()+".total");
+        copyAttributes(id, valueObserver);
+
+        return new OpenTelemetryFunctionTimer<T>(id, obj, countFunction, totalTimeFunction, totalTimeFunctionUnit, getBaseTimeUnit(),
+                countObserver.build(), valueObserver.build());
+    }
+
+    @Override
+    protected <T> FunctionCounter newFunctionCounter(Meter.Id id, T obj, ToDoubleFunction<T> countFunction) {
+        DoubleSumObserver.Builder builder = otelMeter.doubleSumObserverBuilder(id.getName());
+        copyAttributes(id, builder);
+
+        return new OpenTelemetryFunctionCounter<T>(id, obj, countFunction, builder.build());
+    }
+
+    @Override
+    protected TimeUnit getBaseTimeUnit() {
+        return TimeUnit.SECONDS;
+    }
+
+    @Override
+    protected DistributionStatisticConfig defaultHistogramConfig() {
+        return DistributionStatisticConfig.NONE;
+    }
+
+    private void onMeterRemoved(Meter meter) {
+        // .. ?
+    }
+
+    private void copyAttributes(Meter.Id id, AsynchronousInstrument.Builder builder) {
+        builder.setConstantLabels(labelsFrom(id));
+        if ( id.getDescription() != null ) {
+            builder.setDescription(id.getDescription());
+        }
+    }
+
+    private void copyAttributes(Meter.Id id, SynchronousInstrument.Builder builder) {
+        builder.setConstantLabels(labelsFrom(id));
```

Comment:
just FYI, constant labels have been removed from the API, and will be gone with the 0.10.0 release. They may come back in the future, but we removed them for now.

Reply:
ok. so that means I have to hold/defer applying the labels here.. and do them where you noticed the emptyLabels thing was happening instead. thanks